src/tool/hpcrun/gpu/level0/level0-command-list-context-map.c

```diff
@@ -0,0 +1,131 @@
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// * Redistributions of source code must retain the above copyright
+//   notice, this list of conditions and the following disclaimer.
+//
+// * Redistributions in binary form must reproduce the above copyright
+//   notice, this list of conditions and the following disclaimer in the
+//   documentation and/or other materials provided with the distribution.
+//
+// * Neither the name of Rice University (RICE) nor the names of its
+//   contributors may be used to endorse or promote products derived from
+//   this software without specific prior written permission.
+//
+// This software is provided by RICE and contributors "as is" and any
+// express or implied warranties, including, but not limited to, the
+// implied warranties of merchantability and fitness for a particular
+// purpose are disclaimed. In no event shall RICE or contributors be
+// liable for any direct, indirect, incidental, special, exemplary, or
+// consequential damages (including, but not limited to, procurement of
+// substitute goods or services; loss of use, data, or profits; or
+// business interruption) however caused and on any theory of liability,
+// whether in contract, strict liability, or tort (including negligence
+// or otherwise) arising in any way out of the use of this software, even
+// if advised of the possibility of such damage.
+//
+// ******************************************************* EndRiceCopyright *
+
+//*****************************************************************************
+// system includes
+//*****************************************************************************
+
+#include <assert.h>
+#include <string.h>
+
+//*****************************************************************************
+// local includes
+//*****************************************************************************
+
+#include "level0-command-list-context-map.h"
+#include "lib/prof-lean/spinlock.h"
+
+//*****************************************************************************
+// macros
+//*****************************************************************************
+
+#define DEBUG 0
+
+#include "hpcrun/gpu/gpu-print.h"
+
+//******************************************************************************
+// local data
+//******************************************************************************
+
+static level0_handle_map_entry_t *commandlist_context_map_root = NULL;
+
+static level0_handle_map_entry_t *commandlist_context_free_list = NULL;
+
+static spinlock_t commandlist_context_map_lock = SPINLOCK_UNLOCKED;
+
+//*****************************************************************************
+// interface operations
+//*****************************************************************************
+
+ze_context_handle_t
+level0_commandlist_context_map_lookup
```

Comment:
having this use mutual exclusion seems OK for now. I think it  should be marked with a FIXME to reconsider mutual exclusion at a later time.

Reply:
Added TODO comments.